File: Lernwerk/Services/LLM/OpenAICompatibleClient.swift
import Foundation

/// Chat Completions client for OpenRouter and NVIDIA NIM, which both speak the OpenAI wire format.
struct OpenAICompatibleClient: LLMClient {
    static let ocrEngine = "mistral-ocr"
    static let imageHint = " – Falls das Modell keine Bilder versteht: In den Einstellungen „Bilder mitschicken“ ausschalten."

    var provider: LLMProvider
    var apiKey: String
    var model: String
    var sendsImages: Bool
    /// Tried in order when the chosen model is overloaded, times out or no longer exists.
    var fallbackModels: [String] = []
    var session: URLSession = .shared

    var capabilities: LLMCapabilities {
        LLMCapabilities(
            acceptsImages: sendsImages,
            documentHandling: provider == .openRouter ? .providerOCR : .textOnly
        )
    }

    func complete(_ request: LLMRequest) async throws -> LLMResponse {
        var lastError: LLMError = .invalidResponse
        for candidate in [model] + fallbackModels.filter({ $0 != model }) {
            do {
                var response = try await complete(request, model: candidate)
                response.model = response.model ?? candidate
                return response
            } catch let error as LLMError where error.isModelUnavailable {
                lastError = error
            }
        }
        throw lastError
    }

    private func complete(_ request: LLMRequest, model: String) async throws -> LLMResponse {
        guard let url = provider.chatCompletionsURL else {
            throw LLMError.invalidResponse
        }
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.timeoutInterval = request.purpose.timeout
        urlRequest.setValue("application/json", forHTTPHeaderField: "content-type")
        urlRequest.setValue("Bearer \(apiKey)", forHTTPHeaderField: "authorization")
        if provider == .openRouter {
            urlRequest.setValue("Lernwerk", forHTTPHeaderField: "X-Title")
        }

        var body = Self.body(for: request, model: model, provider: provider, sendsImages: sendsImages)
        urlRequest.httpBody = try JSONSerialization.data(withJSONObject: body)
        var (data, status) = try await session.llmData(for: urlRequest)

        // Not every NIM model's chat template knows the thinking switch; retry once without it.
        if status == 400 || status == 422, body["chat_template_kwargs"] != nil {
            body["chat_template_kwargs"] = nil
            urlRequest.httpBody = try JSONSerialization.data(withJSONObject: body)
            (data, status) = try await session.llmData(for: urlRequest)
        }

        return try Self.parse(
            data: data,
            status: status,
            expectsJSON: request.jsonSchema != nil,
            sentImages: Self.containsImages(request, sendsImages: sendsImages)
        )
    }

    static func body(for request: LLMRequest, model: String, provider: LLMProvider, sendsImages: Bool) -> [String: Any] {
        var messages: [[String: Any]] = [["role": "system", "content": systemPrompt(for: request)]]
        messages += request.messages.map { encodeMessage($0, provider: provider, sendsImages: sendsImages) }

        var body: [String: Any] = [
            "model": model,
            "max_tokens": request.maxTokens,
            "messages": messages,
        ]
        let hasPDF = request.messages.contains { message in
            message.content.contains { if case .pdf = $0 { return true } else { return false } }
        }
        if wantsFastAnswer(request) {
            switch provider {
            case .nvidia:
                // GLM/Qwen templates read enable_thinking, Kimi reads thinking; unknown keys are ignored.
                let thinkingOff: [String: Any] = ["enable_thinking": false, "thinking": false]
                body["chat_template_kwargs"] = thinkingOff
            case .openRouter:
                let reasoning: [String: Any] = ["effort": "low"]
                body["reasoning"] = reasoning
            case .google:
                // Gemini 3 cannot switch thinking off; "low" is the shortest it allows.
                body["reasoning_effort"] = "low"
            case .anthropic:
                break
            }
        }
        if hasPDF, provider == .openRouter {
            let parser: [String: Any] = ["id": "file-parser", "pdf": ["engine": ocrEngine]]
            body["plugins"] = [parser]
        }
        return body
    }

    /// All suggested models reason before answering, which takes minutes on free tiers; only the study plan needs that depth.
    static func wantsFastAnswer(_ request: LLMRequest) -> Bool {
        !request.purpose.needsDepth
    }

    /// Not every hosted model enforces JSON schemas (OpenRouter rejects the request instead), so the schema goes into the prompt.
    static func systemPrompt(for request: LLMRequest) -> String {
        guard let schema = request.jsonSchema,
              let data = try? JSONSerialization.data(withJSONObject: schema, options: [.sortedKeys]),
              let schemaText = String(data: data, encoding: .utf8)
        else {
            return request.system
        }
        return """
        \(request.system)

        Output format: reply with exactly one JSON object that validates against this JSON schema. \
        Output only the JSON object, without code fences or explanations.
        <json_schema>
        \(schemaText)
        </json_schema>
        """
    }

    static func encodeMessage(_ message: LLMMessage, provider: LLMProvider, sendsImages: Bool) -> [String: Any] {
        let parts = message.content.compactMap { encodeContent($0, provider: provider, sendsImages: sendsImages) }
        let isTextOnly = parts.allSatisfy { $0["type"] as? String == "text" }
        if message.role == .assistant || isTextOnly {
            let text = parts.compactMap { $0["text"] as? String }.joined(separator: "\n\n")
            return ["role": message.role.rawValue, "content": text]
        }
        return ["role": message.role.rawValue, "content": parts]
    }

    static func encodeContent(_ content: LLMContent, provider: LLMProvider, sendsImages: Bool) -> [String: Any]? {
        switch content {
        case let .text(text):
            return ["type": "text", "text": text]
        case let .image(jpeg):
            guard sendsImages else { return nil }
            let data = provider.maxImageBytes.map { ImageCompressor.jpeg(jpeg, maxBytes: $0) } ?? jpeg
            let imageURL: [String: Any] = ["url": "data:image/jpeg;base64," + data.base64EncodedString()]
            return ["type": "image_url", "image_url": imageURL]
        case let .pdf(data):
            let file: [String: Any] = [
                "filename": "material.pdf",
                "file_data": "data:application/pdf;base64," + data.base64EncodedString(),
            ]
            return ["type": "file", "file": file]
        }
    }

    static func containsImages(_ request: LLMRequest, sendsImages: Bool) -> Bool {
        sendsImages && request.messages.contains { message in
            message.content.contains { if case .image = $0 { return true } else { return false } }
        }
    }

    static func parse(data: Data, status: Int, expectsJSON: Bool, sentImages: Bool) throws -> LLMResponse {
        // Gemini's compatibility endpoint wraps errors in a one-element array.
        let parsed = try? JSONSerialization.jsonObject(with: data)
        let object = (parsed as? [String: Any]) ?? (parsed as? [[String: Any]])?.first
        let apiError = object?["error"] as? [String: Any]

        if !(200..<300).contains(status) || apiError != nil {
            let code = (apiError?["code"] as? Int) ?? status
            let message = (apiError?["message"] as? String)
                ?? (object?["detail"] as? String)
                ?? (object?["title"] as? String)
                ?? String(decoding: data, as: UTF8.self)
            switch code {
            case 401, 403:
                throw LLMError.invalidAPIKey
            case 402:
                throw LLMError.paymentRequired(message)
            case 429:
                throw LLMError.rateLimited(message)
            case 502, 503, 504:
                throw LLMError.overloaded(status: code)
            case 400 where sentImages, 422 where sentImages:
                throw LLMError.http(status: code, message: message + imageHint)
            default:
                throw LLMError.http(status: code, message: message)
            }
        }

        guard let object,
              let choices = object["choices"] as? [[String: Any]],
              let choice = choices.first
        else {
            throw LLMError.invalidResponse
        }

        let finishReason = choice["finish_reason"] as? String
        if finishReason == "content_filter" {
            throw LLMError.refusal
        }

        let message = choice["message"] as? [String: Any]
        let text = ModelText.removingReasoning(contentText(message?["content"]))
        if finishReason == "length", expectsJSON || text.isEmpty {
            throw LLMError.truncated
        }
        guard !text.isEmpty else {
            throw LLMError.invalidResponse
        }
        return LLMResponse(text: text, stopReason: finishReason, model: object["model"] as? String)
    }

    private static func contentText(_ content: Any?) -> String {
        if let text = content as? String {
            return text
        }
        if let parts = content as? [[String: Any]] {
            return parts.compactMap { $0["text"] as? String }.joined()
        }
        return ""
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/llm/OpenAiCompatibleClient.kt
package de.maxifrz.lernwerk.llm

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.util.Base64

/** Chat Completions client for OpenRouter and NVIDIA NIM, which both speak the OpenAI wire format. */
class OpenAiCompatibleClient(
    private val provider: LlmProvider,
    private val apiKey: String,
    private val model: String,
    private val sendsImages: Boolean,
    private val transport: HttpTransport,
    /** Tried in order when the chosen model is overloaded, times out or no longer exists. */
    private val fallbackModels: List<String> = emptyList(),
    /** Shrinks a JPEG below the given size; the Android implementation lives outside the pure logic. */
    private val compressImage: (ByteArray, Int) -> ByteArray = { data, _ -> data },
) : LlmClient {
    override val capabilities = LlmCapabilities(
        acceptsImages = sendsImages,
        documentHandling = if (provider == LlmProvider.OPEN_ROUTER) DocumentHandling.PROVIDER_OCR else DocumentHandling.TEXT_ONLY,
    )

    override suspend fun complete(request: LlmRequest): LlmResponse {
        var lastError: LlmError = LlmError.InvalidResponse
        for (candidate in listOf(model) + fallbackModels.filter { it != model }) {
            try {
                val response = complete(request, candidate)
                return response.copy(model = response.model ?: candidate)
            } catch (error: LlmError) {
                if (!error.isModelUnavailable) throw error
                lastError = error
            }
        }
        throw lastError
    }

    private suspend fun complete(request: LlmRequest, model: String): LlmResponse {
        val url = provider.chatCompletionsUrl ?: throw LlmError.InvalidResponse
        val headers = buildMap {
            put("content-type", "application/json")
            put("authorization", "Bearer $apiKey")
            if (provider == LlmProvider.OPEN_ROUTER) put("X-Title", "Lernwerk")
        }
        val timeout = request.purpose.timeoutSeconds
        var body = body(request, model, provider, sendsImages, compressImage)
        var result = transport.post(url, headers, body.toString(), timeout)

        // Not every NIM model's chat template knows the thinking switch; retry once without it.
        if ((result.status == 400 || result.status == 422) && body.containsKey("chat_template_kwargs")) {
            body = JsonObject(body - "chat_template_kwargs")
            result = transport.post(url, headers, body.toString(), timeout)
        }

        return parse(
            result.body,
            result.status,
            expectsJson = request.jsonSchema != null,
            sentImages = containsImages(request, sendsImages),
        )
    }

    companion object {
        const val OCR_ENGINE = "mistral-ocr"
        const val IMAGE_HINT = " – Falls das Modell keine Bilder versteht: In den Einstellungen „Bilder mitschicken“ ausschalten."

        fun body(
            request: LlmRequest,
            model: String,
            provider: LlmProvider,
            sendsImages: Boolean,
            compressImage: (ByteArray, Int) -> ByteArray = { data, _ -> data },
        ): JsonObject = buildJsonObject {
            put("model", model)
            put("max_tokens", request.maxTokens)
            put("messages", buildJsonArray {
                add(buildJsonObject {
                    put("role", "system")
                    put("content", systemPrompt(request))
                })
                request.messages.forEach { add(encodeMessage(it, provider, sendsImages, compressImage)) }
            })
            if (wantsFastAnswer(request)) {
                when (provider) {
                    // GLM/Qwen templates read enable_thinking, Kimi reads thinking; unknown keys are ignored.
                    LlmProvider.NVIDIA -> put("chat_template_kwargs", buildJsonObject {
                        put("enable_thinking", false)
                        put("thinking", false)
                    })
                    LlmProvider.OPEN_ROUTER -> put("reasoning", buildJsonObject { put("effort", "low") })
                    // Gemini 3 cannot switch thinking off; "low" is the shortest it allows.
                    LlmProvider.GOOGLE -> put("reasoning_effort", "low")
                    LlmProvider.ANTHROPIC -> Unit
                }
            }
            val hasPdf = request.messages.any { message -> message.content.any { it is LlmContent.Pdf } }
            if (hasPdf && provider == LlmProvider.OPEN_ROUTER) {
                put("plugins", buildJsonArray {
                    add(buildJsonObject {
                        put("id", "file-parser")
                        put("pdf", buildJsonObject { put("engine", OCR_ENGINE) })
                    })
                })
            }
        }

        /** All suggested models reason before answering, which takes minutes on free tiers; only the study plan needs that depth. */
        fun wantsFastAnswer(request: LlmRequest) = !request.purpose.needsDepth

        /** Not every hosted model enforces JSON schemas (OpenRouter rejects the request instead), so the schema goes into the prompt. */
        fun systemPrompt(request: LlmRequest): String {
            val schema = request.jsonSchema ?: return request.system
            return listOf(
                request.system,
                "",
                "Output format: reply with exactly one JSON object that validates against this JSON schema. " +
                    "Output only the JSON object, without code fences or explanations.",
                "<json_schema>",
                schema.toString(),
                "</json_schema>",
            ).joinToString("\n")
        }

        fun encodeMessage(
            message: LlmMessage,
            provider: LlmProvider,
            sendsImages: Boolean,
            compressImage: (ByteArray, Int) -> ByteArray = { data, _ -> data },
        ): JsonObject {
            val parts = message.content.mapNotNull { encodeContent(it, provider, sendsImages, compressImage) }
            val isTextOnly = parts.all { it.string("type") == "text" }
            return buildJsonObject {
                put("role", message.role.wire)
                if (message.role == LlmRole.ASSISTANT || isTextOnly) {
                    put("content", parts.mapNotNull { it.string("text") }.joinToString("\n\n"))
                } else {
                    put("content", JsonArray(parts))
                }
            }
        }

        fun encodeContent(
            content: LlmContent,
            provider: LlmProvider,
            sendsImages: Boolean,
            compressImage: (ByteArray, Int) -> ByteArray = { data, _ -> data },
        ): JsonObject? = when (content) {
            is LlmContent.Text -> buildJsonObject {
                put("type", "text")
                put("text", content.text)
            }
            is LlmContent.Image -> if (!sendsImages) null else {
                val data = provider.maxImageBytes?.let { compressImage(content.jpeg, it) } ?: content.jpeg
                buildJsonObject {
                    put("type", "image_url")
                    put("image_url", buildJsonObject {
                        put("url", "data:image/jpeg;base64," + Base64.getEncoder().encodeToString(data))
                    })
                }
            }
            is LlmContent.Pdf -> buildJsonObject {
                put("type", "file")
                put("file", buildJsonObject {
                    put("filename", "material.pdf")
                    put("file_data", "data:application/pdf;base64," + Base64.getEncoder().encodeToString(content.data))
                })
            }
        }

        fun containsImages(request: LlmRequest, sendsImages: Boolean) =
            sendsImages && request.messages.any { message -> message.content.any { it is LlmContent.Image } }

        fun parse(body: String, status: Int, expectsJson: Boolean, sentImages: Boolean): LlmResponse {
            // Gemini's compatibility endpoint wraps errors in a one-element array.
            val obj = parseObject(body) ?: runCatching {
                (kotlinx.serialization.json.Json.parseToJsonElement(body) as? JsonArray)?.firstOrNull() as? JsonObject
            }.getOrNull()
            val apiError = obj?.obj("error")

            if (status !in 200..299 || apiError != null) {
                val code = apiError?.get("code").intOrNull() ?: status
                val message = apiError?.string("message") ?: obj?.string("detail") ?: obj?.string("title") ?: body
                throw when {
                    code == 401 || code == 403 -> LlmError.InvalidApiKey
                    code == 402 -> LlmError.PaymentRequired(message)
                    code == 429 -> LlmError.RateLimited(message)
                    code == 502 || code == 503 || code == 504 -> LlmError.Overloaded(code)
                    (code == 400 || code == 422) && sentImages -> LlmError.Http(code, message + IMAGE_HINT)
                    else -> LlmError.Http(code, message)
                }
            }

            val choice = (obj?.get("choices") as? JsonArray)?.firstOrNull() as? JsonObject
                ?: throw LlmError.InvalidResponse

            val finishReason = choice.string("finish_reason")
            if (finishReason == "content_filter") throw LlmError.Refusal

            val text = ModelText.removingReasoning(contentText(choice.obj("message")?.get("content")))
            if (finishReason == "length" && (expectsJson || text.isEmpty())) throw LlmError.Truncated
            if (text.isEmpty()) throw LlmError.InvalidResponse
            return LlmResponse(text, finishReason, obj.string("model"))
        }

        private fun contentText(content: kotlinx.serialization.json.JsonElement?): String = when (content) {
            is JsonPrimitive -> if (content.isString) content.content else ""
            is JsonArray -> content.mapNotNull { (it as? JsonObject)?.string("text") }.joinToString("")
            else -> ""
        }
    }
}
